keps/sig-apps/2232-jobs-stopped/README.md

```diff
@@ -0,0 +1,558 @@
+18 nodes before scheduling the Pods of C.
+
+If the Job specs had a `stopped` field set to true at the time of creation of
+the object, the controller-manager could use this information and hold off from
+creating the Pods of this Job. This field could be automatically set to true at
+creation with a mutating webhook and then be modified by an external controller
+once cluster resources become available:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: openmpi
+spec:
+  stopped: true  # added with a mutating webhook
+  parallelism: 4
+  template:
+    spec:
+      containers:
+      - name: foo
+        image: bar
+        command: ["baz"]
+```
+
+[Open MPI]: https://www.open-mpi.org/
+
+#### Story 2
+
+<!-- An example to motivate stopping active Jobs? -->
+
+### Notes/Constraints/Caveats
+
+While creating a stopped Job is well-understood, stopping an active Job
+requires some careful thought. Assuming that the semantics of stopping a Job
+are similar to that of a Deployment, the controller manager must delete all
+active Pods when a Job is stopped. Pods that successfully completed will still
+count towards the Job's completions even if the Job is stopped.
+
+Another interesting point of note is that it is important that system
+administrators restricts users from accidentally modifying this field manually
+in a Job managed by a higher-level orchestrator. This can be done with a
+mutating webhook, for example.
+
+It is intended that this feature *not* add config toil for the end-user.
+Through policies in the higher-level orchestrator, one can enforce that all
+Jobs in a cluster automatically get created as stopped Jobs.
+
+Exactly how and when a Job is resumed is left to the orchestrator (e.g. cluster
+scaling succeeded, quotas are relaxed, or the Job is at the head of the queue).
+Note that Pods created by the Job controller are still scheduled to nodes by
+the scheduler.
+
+### Risks and Mitigations
+
+1. Stopping an active Job will delete *all* active pods.
+
+2. It is important to prevent users from interfering with the proposed field
+   manually in jobs managed by a higher-level orchestrator.
+
+## Design Details
+
+We propose adding a `stopped` field to the JobSpec API:
+
+```go
+type JobSpec struct {
+	// Stopped specifies whether the job controller should create Pods or not. If
+	// a Job is created in the stopped state, no Pods are created by the job
+	// controller. If a Job enters the stopped state (i.e. the flag goes from
+	// false to true), the job controller will delete all Pods associated with
+	// this job. This field is generally used by higher-level orchestrators that
+	// need more control over when the Pods of a Job are created.
+	Stopped *bool
+
+	...
+}
+```
+
+As described in the comment, when the boolean is set to true, the
+controller-manager will abstain from creating Pods even if the Job's
+parallelism requirements aren't met. If the Job is already active and is
```

Comment:
Add more detail on how stop is handled when there are active Pods.

Things to think about:
- will Pods have forceful termination?
- If the Pods receive a termination signal, and they exit with a 0 exit code, wouldn't kubelet mark them as succeeded? That would cause the Job to increase it's completions counter or finish (if `.spec.completions` is not set). Maybe some of this scenarios are desirable in the case of preemption, and users are expected to create a new Job.

Reply:
@adtac can you describe the parallelism=0 semantics? we may be ok if we simply fallback to that behavior.

Reply:
It just doesn't create new Pods. The existing Pods continue running.

Reply:
No, decreasing parallelism will delete pods:                                                    

https://github.com/kubernetes/kubernetes/blob/faaae714274c2fd0eb6ce71a77d54106b7689e3d/pkg/     controller/job/job_controller.go#L702

which deletes the pod:

https://github.com/kubernetes/kubernetes/blob/faaae714274c2fd0eb6ce71a77d54106b7689e3d/pkg/     controller/controller_utils.go#L600

I'll post an update with the exact exit code, but I don't think it's 0.

Reply:
I just verified that the job's pods that get terminated do not count towards the completions. In fact, I can't even query the pod with `kubectl get`, so it looks like all references to it are removed.

Reply:
I guess this makes sense. It's an explicit deletion, even if it takes some time for the actual deletion to happen.

Reply:
Add the detail of how Pod termination is handled to the KEP.

However, I still have one doubt. I assume the process goes like follows:
1. Controller calls DELETE on the API
2. kube-apiserver adds the deletionTimestamp
3. Kubelet sends the termination signal
4. Containers terminate 
5. kubelet marks the pod as finished or removes the Pod?

If kubelet first marks the Pod as finished (Failed or Succeeded) and then deletes the Pod, there could be a time where the controller counts the Pod towards the Job status.
OTOH, if kubelet directly removes the Pod, then there is no chance for the Job controller to use this Pod towards status.

Reply:
can you resolve this please.

Reply:
>Add the detail of how Pod termination is handled to the KEP

Done.

>kubelet marks the pod as finished or removes the Pod?

For others: we discussed this offline and concluded that the Kubelet will never mark these pods as finished:

https://github.com/kubernetes/kubernetes/blob/9da11e294fae8e26330d5c0ed5dff61cda178d77/staging/src/k8s.io/api/core/v1/types.go#L2509

Pods terminated by the system will always be marked as failed, so they will not count towards completions.